pkg/resource/provider/component_provider.go

```diff
@@ -0,0 +1,162 @@
+// Copyright 2016-2021, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package provider
```

Comment:
Any reason not to use [`plugin.providerServer`](https://github.com/pulumi/pulumi/blob/master/sdk/go/common/resource/plugin/provider_server.go)? Looks like their functionality is similar.

Reply:
The `Construct` adapter func in this PR adapts the raw gRPC interface to/from the Pulumi Go SDK model.

https://github.com/pulumi/pulumi/blob/a8ea363df970c58709097b3ab28e82afc5a0d9ce/sdk/proto/go/provider.pb.go#L2047-L2048

So the provider in this file implements the raw gRPC interface.

To use `plugin.providerServer`, `Construct` would have to be an adapter of `plugin.providerServer`'s interface:

https://github.com/pulumi/pulumi/blob/664776efe85ca4c6a8c9cd95fa2fbfa75f2dbf03/sdk/go/common/resource/plugin/provider.go#L78-L80

Should `Construct` be an adapter for `plugin.providerServer` instead? Or should we have adapters for both the raw gRPC interface **and** `plugin.providerServer`?

The reason I made `Construct` an adapter for the raw gRPC interface is because 1) our existing [provider boilerplate repo](https://github.com/pulumi/pulumi-provider-boilerplate) implements a provider using the raw gRPC interface rather than `plugin.providerServer` and 2) our own native providers also implement the raw gRPC interface over `plugin.providerServer`.

Reply:
> Should Construct be an adapter for plugin.providerServer instead? Or should we have adapters for both the raw gRPC interface and plugin.providerServer?
> 
> The reason I made Construct an adapter for the raw gRPC interface is because 1) our existing provider boilerplate repo implements a provider using the raw gRPC interface rather than plugin.providerServer and 2) our own native providers also implement the raw gRPC interface over plugin.providerServer.

IMO we should adapt from `plugin.providerServer`, but I think that we can move forward with what you have. In the fullness of time, I'd love for all of our Go providers to be sitting on top of `providerServer`.